sympy/sets/fancysets.py

```diff
@@ -49,18 +48,27 @@ def _contains(self, other):
         return tfn[other.is_rational]
 
     def __iter__(self):
+        r""" Enumerate rational numbers.
+
+        References
+        ==========
+
+        .. [1] Aimeric Malter, Dierk Schleicher, Don Zagier, New Looks at Old Number Theory,
+               The American Mathematical Monthly, Vol. 120, No. 3 (2013), pp. 243-264
+               https://doi.org/10.4169/amer.math.monthly.120.03.243
+
+        """
         yield S.Zero
         yield S.One
         yield S.NegativeOne
-        d = 2
-        while True:
-            for n in range(d):
-                if igcd(n, d) == 1:
-                    yield Rational(n, d)
-                    yield Rational(d, n)
-                    yield Rational(-n, d)
-                    yield Rational(-d, n)
-            d += 1
+        for q in count(2):
+            p = 1
+            for _ in range(1 << (q-2)):
+                yield Rational(p, q)
+                yield Rational(q, p)
+                yield Rational(-p, q)
+                yield Rational(-q, p)
+                p, q = q, q*(1 + ((p // q) << 1)) - p
```

Comment:
I think that you may clarify my understanding of the idea behind the formula:

First of all, you may be taking advantage of the symmetry of tree to calculate only left half of the tree, and use

yield Rational(p, q)
yield Rational(q, p)

Secondly, the expression

q, q*(1 + ((p // q) << 1)) - p

Can be understood easily, to be same formula appearing in the article, I divide q in both sides, right?

Reply:
Each call to Rational involves a gcd calculation. It would potentially be more efficient to use:
```
r1 = Rational(p, q)
r2 = 1/r1
yield from [r1, r2, -r1, -r2]
```

Reply:
The most faithful implementation is as follows

```python
def rationals_iter():
    x = S.Zero
    while True:
        x = Rational(1, 1 + 2*x.floor() - x)
        yield x
```

When `x=p/q`, we can transform the expression as follows

```math
\frac{1}{2\left\lfloor x \right\rfloor + 1 - x} = \frac{q}{2q\left\lfloor p / q \right\rfloor + q - p}
```

> I think that you may clarify my understanding of the idea behind the formula

Your understanding is correct.